web3/middleware/validation.py

```diff
@@ -1,19 +1,91 @@
 from cytoolz import (
+    assoc,
+    complement,
     compose,
     curry,
     dissoc,
 )
 from eth_utils.curried import (
-    apply_formatter_at_index,
+    apply_formatter_if,
     apply_formatters_to_dict,
+    is_null,
+)
+from eth_utils.decorators import (
+    return_arg_type,
+)
+from eth_utils.functional import (
+    to_dict,
+)
+from hexbytes import (
+    HexBytes,
 )
 
 from web3.exceptions import (
     ValidationError,
 )
 
+is_not_null = complement(is_null)
+
 
+# TODO: move to utils
+def construct_validation_middleware(request_validators=None, result_validators=None):
+    if request_validators is None:
+        request_validators = {}
+    if result_validators is None:
+        result_validators = {}
+
+    def formatter_middleware(make_request, web3):
+        def middleware(method, params):
+            if method in request_validators:
+                formatter = request_validators[method]
+                formatted_params = formatter(web3, params)
+                response = make_request(method, formatted_params)
+            else:
+                response = make_request(method, params)
+            if 'result' in response and method in result_validators:
+                formatter = result_validators[method]
+                formatted_response = assoc(
+                    response,
+                    'result',
+                    formatter(response['result']),
+                )
+                return formatted_response
+            else:
+                return response
+        return middleware
+    return formatter_middleware
+
+
+# TODO: move to utils
 @curry
+@return_arg_type(3)
+def apply_validator_at_index(validator, at_index, web3, value):
+    if at_index + 1 > len(value):
+        raise IndexError(
+            "Not enough values in iterable to apply formatter.  Got: {0}. "
+            "Need: {1}".format(len(value), at_index + 1)
+        )
+    for index, item in enumerate(value):
+        if index == at_index:
+            yield validator(web3, item)
+        else:
+            yield item
+
+
+# TODO: move to utils
+@curry
+@to_dict
+def apply_validator_to_dict(validators, web3, value):
+    for key, item in value.items():
+        if key in validators:
+            try:
+                yield key, validators[key](web3, item)
+            except (TypeError, ValueError) as exc:
```

Comment:
need to remove this check as vaildators will raise their own exceptions on failure.